infra/src/main/kotlin/fi/fta/geoviite/infra/publication/PublicationDao.kt

```diff
@@ -394,6 +398,431 @@ class PublicationDao(
+                left join common.switch_structure switch_structure
+                          on switch_structure.id = switch_version.switch_structure_id
+                left join common.switch_owner
+                          on switch_owner.id = switch_version.owner_id
+                left join geometry.switch gs
+                          on switch_version.geometry_switch_id = gs.id
+                left join geometry.plan plan
+                          on gs.plan_id = plan.id
+                left join layout.switch_version old_switch_version
+                          on old_switch_version.id = switch_version.id and old_switch_version.version = switch_version.version - 1 and old_switch_version.draft = false
+                left join common.switch_structure old_switch_structure
+                          on old_switch_structure.id = old_switch_version.switch_structure_id
+                left join common.switch_owner old_switch_owner
+                          on old_switch_owner.id = old_switch_version.owner_id
+                left join geometry.switch old_gs
+                          on old_switch_version.geometry_switch_id = old_gs.id
+                left join geometry.plan old_plan
+                          on old_gs.plan_id = old_plan.id
+                left join joints
+                          on joints.switch_id = switch.switch_id
+                left join location_tracks
+                          on location_tracks.switch_id = switch.switch_id
+              where publication_id = :publication_id
+        """.trimIndent()
+
+        return jdbcTemplate.query(sql, mapOf("publication_id" to publicationId.intValue)) { rs, _ ->
+            val presentationJointNumber = rs.getInt("presentation_joint_number").let(::JointNumber)
+            val trackNumbers = rs.getListOrNull<Int>("track_number_ids")?.map { IntId<TrackLayoutTrackNumber>(it) } ?: emptyList()
+            val locationTracks = rs.getListOrNull<Int>("location_track_ids")?.map { IntId<LocationTrack>(it) } ?: emptyList()
+            val jointNumbers = rs.getListOrNull<Int>("joint_numbers") ?: emptyList()
+            val removed = rs.getListOrNull<Boolean>("removed") ?: emptyList()
+            val points = rs.getListOrNull<Double>("points_x")?.zip(rs.getList<Double>("points_y"))?.map { Point(it.first, it.second) } ?: emptyList()
+            val addresses = rs.getListOrNull<String>("addresses")?.map(::TrackMeter) ?: emptyList()
+            val joints = jointNumbers.indices.map {  index ->
+                PublicationSwitchJoint(
+                    jointNumber = JointNumber(jointNumbers[index]),
+                    locationTrackId = locationTracks[index],
+                    removed = removed[index],
+                    point = points[index],
+                    address = addresses[index],
+                    trackNumberId = trackNumbers[index],
+                )
+            }.filter { joint -> joint.jointNumber == presentationJointNumber }
```

Comment:
Filteröinnin voi varmaan tehdä jo SQL-puolella, jolloin indeksillä pyörittely jäisi täällä pois

Reply:
Lähdin pöyhimään tätä, mutta tämä ei olekaan kauhean simppeliä. Syynä se, että käytännössä filtteröinti pitäisi kohdistaa ei niinkään `joints`-riveihin, vaan niiden sisäisiin taulukkoihin, ja tämä sama pitäisi tehdä jokaiselle taulukolle tuon `joint_numbers`-taulukon perusteella. Jätän siis tämän toistaiseksi näin, mutta jos joku muu haluaa palata tähän myöhemmin, niin tämä SQL on all theirs